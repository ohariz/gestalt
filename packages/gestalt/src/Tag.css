.button {
  composes: noBorder from "./Borders.css";
  composes: paddingX2 from "./boxWhitespace.css";
  composes: paddingY3 from "./boxWhitespace.css";
  composes: pointer from "./Cursor.css";
}

html[dir="rtl"] .button {
  border-radius: calc(var(--g-boint) * 2) 0 0 calc(var(--g-boint) * 2);
}

html:not([dir="rtl"]) .button {
  border-radius: 0 calc(var(--g-boint) * 2) calc(var(--g-boint) * 2) 0;
}

.lightGray {
  composes: lightGrayBg from "./Colors.css";
}

.button.lightGray:hover,
.button.lightGray:focus {
  background-color: var(--g-colorGray100Hovered);
}

.button.lightGray:active {
  background-color: var(--g-colorGray100Active);
}

.red {
  composes: redBg from "./Colors.css";
}

.red:hover,
.red:focus {
  background-color: var(--g-colorRed100Hovered);
}

.red:active {
  background-color: var(--g-colorRed100Active);
}
