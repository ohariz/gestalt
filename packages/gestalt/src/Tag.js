// @flow strict
import React, { type Node } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Box from './Box.js';
import Flex from './Flex.js';
import Text from './Text.js';
import Icon from './Icon.js';
import useFocusVisible from './useFocusVisible.js';
import focusStyles from './Focus.css';
import touchableStyles from './Touchable.css';
import styles from './Tag.css';

type Props = {|
  text: string,
  errorMessage?: string,
  ...
    | {|
        disabled: true,
        onRemove?: ({|
          event: SyntheticMouseEvent<HTMLButtonElement>,
        |}) => void,
        removeIconAccessibilityLabel?: string,
      |}
    | {|
        disabled?: false,
        onRemove: ({|
          event: SyntheticMouseEvent<HTMLButtonElement>,
        |}) => void,
        removeIconAccessibilityLabel: string,
      |},
|};

export default function Tag(props: Props): Node {
  const {
    disabled = false,
    errorMessage,
    onRemove = null,
    removeIconAccessibilityLabel = '',
    text,
  } = props;

  const bgColor = errorMessage ? 'red' : 'lightGray';
  let fgColor = 'darkGray';
  if (errorMessage) {
    fgColor = 'white';
  } else if (disabled) {
    fgColor = 'gray';
  }

  const { isFocusVisible } = useFocusVisible();

  const removeIconClasses = classnames(
    styles.button,
    styles[bgColor],
    focusStyles.hideOutline,
    touchableStyles.tapTransition,
    {
      [focusStyles.accessibilityOutline]: isFocusVisible,
    }
  );

  return (
    <Box
      borderStyle={disabled && !errorMessage ? 'sm' : 'none'}
      color={bgColor}
      display="inlineBlock"
      rounding={2}
    >
      <Flex alignItems="center">
        {errorMessage && (
          <Box marginStart={2}>
            <Icon
              accessibilityLabel={errorMessage}
              color={fgColor}
              icon="workflow-status-problem"
            />
          </Box>
        )}
        <Box
          marginStart={2}
          marginEnd={disabled ? 2 : 1}
          paddingY={disabled ? 2 : 0}
        >
          <Text color={fgColor} size="md">
            {text}
          </Text>
        </Box>
        {!disabled && (
          <button className={removeIconClasses} onClick={onRemove} type="button">
            <Icon
              accessibilityLabel={removeIconAccessibilityLabel}
              color={fgColor}
              icon="cancel"
              size={8}
            />
          </button>
        )}
      </Flex>
    </Box>
  );
}

Tag.propTypes = {
  text: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  errorMessage: PropTypes.string,
  onRemove: PropTypes.func,
  removeIconAccessibilityLabel: PropTypes.string,
};
