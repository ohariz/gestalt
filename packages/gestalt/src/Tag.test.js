// @flow strict
import React from 'react';
import { create } from 'react-test-renderer';
import Tag from './Tag.js';

describe('Tag', () => {
  it('renders', () => {
    const tree = create(
      <Tag
        text="New"
        onRemove={() => {}}
        removeIconAccessibilityLabel="Remove"
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a disabled tag', () => {
    const tree = create(<Tag text="New" disabled />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a tag with an error state', () => {
    const tree = create(
      <Tag
        text="New"
        onRemove={() => {}}
        removeIconAccessibilityLabel="Remove"
        errorMessage="Error"
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
