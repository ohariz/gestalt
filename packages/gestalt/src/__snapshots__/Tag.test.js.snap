// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag renders 1`] = `
<div
  aria-disabled={false}
  className="box lightGrayBg rounding2 xsDisplayInlineBlock"
  style={
    Object {
      "height": 32,
    }
  }
>
  <div
    className="Flex rowGap0 itemsCenter xsDirectionRow"
    style={
      Object {
        "height": "100%",
      }
    }
  >
    <div
      className="box marginEnd1 marginStart2"
    >
      <span
        className="Text fontSize2 darkGray alignLeft breakWord fontWeightNormal"
      >
        New
      </span>
    </div>
    <button
      className="button lightGray hideOutline tapTransition accessibilityOutline"
      onClick={[Function]}
      type="button"
    >
      <svg
        aria-hidden={null}
        aria-label="Remove"
        className="icon darkGray iconBlock"
        height={8}
        role="img"
        viewBox="0 0 24 24"
        width={8}
      >
        <path
          d="test-file-stub"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`Tag renders a disabled tag 1`] = `
<div
  aria-disabled={true}
  className="borderColorLightGray box lightGrayBg rounding2 sizeSm solid xsDisplayInlineBlock"
  style={
    Object {
      "height": 32,
    }
  }
>
  <div
    className="Flex rowGap0 itemsCenter xsDirectionRow"
    style={
      Object {
        "height": "100%",
      }
    }
  >
    <div
      className="box marginEnd2 marginStart2"
    >
      <span
        className="Text fontSize2 gray alignLeft breakWord fontWeightNormal"
      >
        New
      </span>
    </div>
  </div>
</div>
`;

exports[`Tag renders a tag with an error state 1`] = `
<div
  aria-disabled={false}
  className="box redBg rounding2 xsDisplayInlineBlock"
  style={
    Object {
      "height": 32,
    }
  }
>
  <div
    className="Flex rowGap0 itemsCenter xsDirectionRow"
    style={
      Object {
        "height": "100%",
      }
    }
  >
    <div
      className="box marginStart2"
    >
      <svg
        aria-hidden={null}
        aria-label="Error"
        className="icon white iconBlock"
        height={12}
        role="img"
        viewBox="0 0 24 24"
        width={12}
      >
        <path
          d="test-file-stub"
        />
      </svg>
    </div>
    <div
      className="box marginEnd1 marginStart2"
    >
      <span
        className="Text fontSize2 white alignLeft breakWord fontWeightNormal"
      >
        New
      </span>
    </div>
    <button
      className="button red hideOutline tapTransition accessibilityOutline"
      onClick={[Function]}
      type="button"
    >
      <svg
        aria-hidden={null}
        aria-label="Remove"
        className="icon white iconBlock"
        height={8}
        role="img"
        viewBox="0 0 24 24"
        width={8}
      >
        <path
          d="test-file-stub"
        />
      </svg>
    </button>
  </div>
</div>
`;
